### ISSUE TITLE
c13/c14 isotope time-series is awkward for 1850 control (convert to streams file)

### ISSUE THREAD

**Issue by ekluzek** [ESCOMP/CTSM#182]
**Erik Kluzek < erik > - 2017-11-10 16:15:08 -0700**
**Bugzilla Id:** [2533](http://bugz.cgd.ucar.edu/show_bug.cgi?id=2533)
**Bugzilla CC:**  cdkoven,  dlawren,  klindsay,  rfisher, 

Currently if you do a 1850 control, the assumed c13/c14 is for present day (delta of 0). You can get around this by asking for the c13/c14 time-series files, and if your start year is 0001, it'll use the 1850 value from the file. It's a little awkward since if you run a control for more than one thousand eight hundred and fifty years, it'll start running through the time-series file. For future scenarios it'll use the last value in the file, which is the most reasonable thing to do.

I think the best way around it, is to change the handling of this file from custom code to using the streams mechanism. This will actually shorten the amount of code needed in CLM as well as give more control, options, and flexibility in it's handling. The control for how the data was handled would be in the namelist for it and more obvious.

I think the sectors can be handled in streams by just representing them as three grid cells. This would also move the description of the data from baked inside the code to an outside grid file that could be changed with new datasets.

Definition of done:
- [x] Create script to convert CMIP7 to streams format
- [x] Include a mesh file for the C14 dataset since it has a latitude dependence
- [x] Move datasets to inputdata
- [x] Change code to use the new files @ekluzek 
- [ ] rimport datasets @slevis up to this point, when we confirm that the file work correctly as streams

**Comment by ekluzek** [comment]
**Erik Kluzek < erik > - 2017-11-10 16:16:48 -0700**

This also means the default for 1850 and transient cases when use_c13/use_c14 are on should be to also turn on the time-series files for these cases. That'll do the right thing most of the time.

**Comment by ekluzek** [comment]
**Keith Lindsay < klindsay > - 2017-11-13 09:38:34 -0700**

Erik,

When I looked at using streams to handle the CMIP6 dataset, it looked to me like the shr code didn't handle the time units in the dataset.

time:units in the dataset is "years since 1850-01-01 0:0:0.0", and shr_string_parseCFtunit in shr_string_mod.F90 has the comment
!  Input string is like "days since 0001-06-15 15:20:45.5 -6:00"
!    - recognizes "days", "hours", "minutes", "seconds"
!    - must have at least yyyy-mm-dd, hh:mm:ss.s is optional
!    - expects a "since" in the string
!    - ignores time zone part


Assuming I'm interpreting the above correctly, it seems like one of the following needs to happen:
1) extend shr code to handle this dataset's time:units
2) change the dataset to have time:units that shr code handles

Option 1) seemed unappealing to me because of the shr code's support for both no_leap and gregorian calendars. It seemed like it might be complicated to implement new functionality in shr code while not being consistent with the existing shr code.

If you go the route of option 2), I'll need to modify POP's code that processes this dataset, assuming that we want to continue having CLM and POP read the same file for this forcing. Please let me know if decide to pursue this path.

Keith

**Comment by ekluzek** [comment]
**Keith Lindsay < klindsay > - 2017-11-13 09:59:10 -0700**

(In reply to Erik Kluzek from comment #1)
> This also means the default for 1850 and transient cases when
> use_c13/use_c14 are on should be to also turn on the time-series files for
> these cases. That'll do the right thing most of the time.

It sounds like this will lead to using the 1850 value for these forcings, in spinup runs, and in the 1850 control runs.

The 1850 values (permil) of Delta14C in the CMIP6 dataset in the different latitude bands are
90S-30S: -5.8
30S-30N: -4
30N-90N: -2.3

Is that what we want to be using in our BGC spinups and in the 1850 control runs? I'm used to using 0 permil for PI spinups, though this does create a jump at 1850.

I can see arguments for using 0 or for using the 1850 values.

I definitely think we should use the same value in the spinups as in the 1850 control runs. I also think we should use the same values in CLM and in POP.

Unfortunately, I don't see a specification for this in the C4MIP GMD paper, Jones et al. 2016.

**Comment by ekluzek** [comment]
**charlie < cdkoven > - 2017-11-13 10:11:23 -0700**

Agreed with Keith that there are arguments for using either 0 or the PI values during spinup and PI controls.  I suppose I'd vote for using the PI values, even though that's not what my original code does, in the interest of avoiding artificial discontinuities.

**Comment by ekluzek** [comment]
**Keith Lindsay < klindsay > - 2017-11-13 10:18:54 -0700**

(In reply to Keith Lindsay from comment #2)

> Option 1) seemed unappealing to me because of the shr code's support for
> both no_leap and gregorian calendars. It seemed like it might be complicated
> to implement new functionality in shr code while not being consistent with
> the existing shr code.

I meant "... while being consistent ...".
The not slipped in from a different chain of thought.

**Comment by ekluzek** [comment]
**David Lawrence < dlawren > - 2017-11-13 10:29:45 -0700**

It makes more sense to me to use 1850 values for spinup and 1850 control. Question: if we were using zero per mil, then can you run with isotopes off during spinup and in the 1850 control? ... and only turn them on for the transient, thereby saving some computer time?

In the end, I think the decision on what to do should come from Charlie and Keith since they have more experience with isotopes.

**Comment by ekluzek** [comment]
**charlie < cdkoven > - 2017-11-13 10:34:51 -0700**

re Dave's Question: using zero or the PR value shouldn't really help either way for either isotope. For 14C, since the spinup problem is equilibration with respect to radioactive decay, you still have to do that (albeit in an accelerated way under AD).  For 13C, the spinup problem there is with respect to isotopic fractionation, so, again, you still have to actually do it.  The atmospheric value just becomes the boundary condition for the system in both cases, and in both cases you need to do the equilibration in AD mode in order to let it propagate rapidly through all the pools.

**Comment by ekluzek** [comment]
**Erik Kluzek < erik > - 2017-11-13 11:11:58 -0700**

(In reply to Keith Lindsay from comment #2)
> Assuming I'm interpreting the above correctly, it seems like one of the
> following needs to happen:
> 1) extend shr code to handle this dataset's time:units
> 2) change the dataset to have time:units that shr code handles
> 
> Option 1) seemed unappealing to me because of the shr code's support for
> both no_leap and gregorian calendars. It seemed like it might be complicated
> to implement new functionality in shr code while not being consistent with
> the existing shr code.
>

I looked into this as well. The meaning is a bit awkward for the Gregorian calendar. And because of this the CF contention warns against using a unit of years. But, it says you define the interval as 365 for no_leap or 365.2425 for Gregorian. So how exact the end result would be depends on the reference date.

It wouldn't be hard to extend the code in cime/src/share/util/shr_string_mod to handle this. But, because of this issue we probably shouldn't. I'm bringing it up as an issue to CSEG, we'll see what feedback they have.

Here's the caution in the CF convention...

http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/build/ch04s04.html

> We recommend that the unit year be used with caution. The Udunits 
> package defines a year to be exactly 365.242198781 days (the interval 
> between 2 successive > passages of the sun through vernal equinox). It 
> is not a calendar year. Udunits> > includes the following definitions for 
> years: a common_year is 365 days, a leap_year is 366 days, a Julian_year 
> is 365.25 days, and a Gregorian_year is 365.2425 days.
> 
> For similar reasons the unit month, which is defined in udunits.dat to be 
> exactly year/12, should also be used with caution.

**Comment by ekluzek** [comment]
**Keith Lindsay < klindsay > - 2017-11-17 15:53:01 -0700**

Regarding using 1850 values or 0 permil for 1850 control runs and spinup, I've sought input from Alex Jahn, who developed the C isotopes in POP, and Bette Otto-Bliesner and Esther Brady, as representatives of the paleo and isotope users. Everyone has either no preference or a slight preference for using 1850 values. So counting those three, Charlie and myself, no one has expressed a strong preference for using 0 permil.

So I think we've reached a consensus to use the 1850 values.

Note that this is independent of switching to shr code.

I do not plan on changing POP to use shr code, unless the dataset's time:units is changed. I will change POP to use the 1850 values in 1850 controls and spinup runs. (This is a little quirky for 14C because of the latitude bands in the dataset, but that's the way it is.)

**Comment by ekluzek** [comment]
Note, that this mechanism was inadvertently turned on in for 1850 control simulations. Since, the default is for 1850 conditions, it really didn't need to be. It only causes a problem if you run for more than 1850 years though. But, there were some simulations that ran for that long.

**Comment by ekluzek** [comment]
I've worked on this more and I want to list the issues with the current implementation. The starting year (1850) is hardcoded in, as well as that the time units will be in years. The number of sectors is hardcoded, as are the width of the bands. It will die if the number of sectors is wrong, but won't detect a change in the latitude band width. The time unit of years is considered a bad unit, because you can't detect a difference between Gregorian or no-leap calendars (like you can for other units). It does check the units, so will die if the starting year is wrong, or time isn't years, or the wrong units are used for C13 or C14. So that's good. But, the user also has no control over what years are used in the file, nor spatial nor temporal interpolation. All of these things come out of the box with a streams implementation. And it's trusted shared code, rather than custom code, so better from a maintenance standpoint. The streams implementation will automatically adjust to changes in both the temporal and spatial description of the file. The current mechanism also can't choose the time-period, which means the same file can't be used for Modern as for 20th Century. And as already shown this implementation has #592.

**Comment by ekluzek** [comment]
One other problem (with the current dataset not necessarily in the code itself) is that you can't do ncview on the file: $DIN_LOC_ROOT/lnd/clm2/isotopes/atm_delta_C14_CMIP6_3x1_global_1850-2015_yearly_v2.0_c171012.nc. This is because of the metadata on the "sector" variable is incorrect for CF. Using flag_* attributes allows you to view the file in ncview.

**Comment by billsacks** [comment]
Blocks #592

**Comment by ekluzek** [comment]
This comes up again as we are bringing in the new CMIP7 Carbon Isotope datasets.

The issue above about time units is resolved in the CMIP7 datasets as they are now in a standard unit of days since 1700-01-01.  The number of sectors for C14 is increased to 4, from 3, and the band widths change as well. But, the limitations of the current code still apply and it would be far better to have the file in a format that can be read in by streams to make this general. It would then be easier to swap in new datasets and have better error checking that's in one place in CESM rather than custom code for JUST these two datasets.

Another problem I realized is that CTSM currently only uses the C13/C14 isotope data for Photosynthesis and uses set C13/C14 ratios throughout the rest of the code.  We undoubtably want it to use the same ratios throughout the code and not just for Photosynthesis.

**Comment by wwieder** [comment]
For these comments, Erik.  

Given it was 8 years ago when this issue last came up, and it may be another 8 years before it happens again, I'd have to mark the conversion to streams as a low priority issue.

I'm not sure I understand the other comment you made related to the use of the isotope data in photosynthesis.  I think this is the only place in the code that we should be reading in the atmospheric 13CO2 and 14CO2 data.  After photosynthesis occur CLM tracks the C13 and C14 ratios as C moves around the ecosystem and (radioactively decays), but we don't have to track down the 13 or 14CO2 back to the atmosphere after it's respired.  That is I **don't** think we want to do this 

>_"We undoubtably want it to use the same ratios throughout the code and not just for Photosynthesis."_

But will defer to @ckoven here to confirm

**Comment by ekluzek** [comment]
> Given it was 8 years ago when this issue last came up, and it may be another 8 years before it happens again, I'd have to mark the conversion to streams as a low priority issue.

Yes, since it changes infrequently it might be OK to largely leave alone.  I just want to mark this as something with technical debt that also makes the model less flexible for science, as well as more vulnerable to bugs because of the lack of error checking.

> 
> I'm not sure I understand the other comment you made related to the use of the isotope data in photosynthesis. I think this is the only place in the code that we should be reading in the atmospheric 13CO2 and 14CO2 data. After photosynthesis occur CLM tracks the C13 and C14 ratios as C moves around the ecosystem and (radioactively decays), but we don't have to track down the 13 or 14CO2 back to the atmosphere after it's respired. That is I **don't** think we want to do this
> 
> > _"We undoubtably want it to use the same ratios throughout the code and not just for Photosynthesis."_
> 
> But will defer to [@ckoven](https://github.com/ckoven) here to confirm

I'll start another issue for this to go into more depth. But, the code sets c13ratio and c14ratio which are hardcoded and described as the pre-industrial atmospheric c13ratio. This is used in places in the model that look like it should use rc13_canair_patch and rc14_canair_patch. For example see this description of c13ration in clm_varcon.F90:

``` fortran
 !!! C13
  real(r8), public, parameter :: preind_atm_del13c = -6.0_r8   ! preindustrial value for atmospheric del13C
  real(r8), public, parameter :: preind_atm_ratio = SHR_CONST_PDB + (preind_atm_del13c * SHR_CONST_PDB)/1000.0_r8  ! 13C/12C
  real(r8), public :: c13ratio = preind_atm_ratio/(1.0_r8+preind_atm_ratio) ! 13C/(12+13)C preind atmosphere
```

Once, the Carbon isotopes are in the plants -- I'm assuming it's handled correctly. It's these use of c13ratio and c14ratio that clearly define themselves as isotope ratio in air that I question. Especially, when above it seems to describe it as fixed to preindustrial.

**Comment by ekluzek** [comment]
To be closer to do this for CMIP7 I'm adding the metadata changes that would be needed for that:

To allow us to use streams on the dataset in the future @slevis-lmwg will do a few things to make it readable:
- [x] Change calendar to "noleap" for time and time_bnds (start a script to do these to put in the CMIP7 repo)
- [x] Add lat/lon and lat_bnds/lon_bnds to define the gridcells (latitude bands for C14 and a global gridcell for C13), see Erik's post below for more info

For the next change wait until ready to switch to streams (else need to change the code):
- Dimensions of data can now be: time,lat,lon 
- Time dimension needs to be changed to unlimited

**Comment by ekluzek** [comment]
Here's a comment about what we think the metadata should look like on the file:

https://github.com/ESCOMP/CTSM/issues/3346#issuecomment-3358408395

In addition to that time needs to be changed into the unlimited dimension.

**Comment by ekluzek** [comment]
Yesterday we decided especially because there are multiple problems with the current approach (notably #592)  that we should do the conversion to a streams file to remove the issues, and make it more robust, extensible and flexible with any bugs primarily being in reused code rather than custom code that is more likely to have problems, and less likely to be easy to debug or notice.

This means @slevis-lmwg will do the work on converting the dataset in order to allow it be used by streams, with creating a script that will go into the CMIP7 inputdata repository (https://github.com/NCAR/CMIP7_inputdata_processing). First draft of script currently here: 
`/fs/cgd/csm/inputdata/lnd/clm2/isotopes/rawdata_copies/workspace/prep_c13_14_streams.sh`

And I'll work on the code changes to use streams in the code. We want to finish this in a week or two.

**Comment by wwieder** [comment]
I think this makes sense, thanks @ekluzek and @slevis-lmwg 

One point of clarification (and sorry to be a pain), but yesterday at the SE meeting Erik you estimated under a week of work for this, but  above you suggested:
 
> We want to finish this in a week or two.

Is there a reason this should take more than a week to wrap up?

**Comment by ekluzek** [comment]
Thanks for asking @wwieder. I was mostly being lax in my statement about a week or two. But, also I think my SE time is less than a week, but with downtime due to coordination and waiting on testing, I could see it be up to 2 weeks to actually get it in with a tag. But, let me work towards that week timeline. There is always the difference between the estimate based on total SE time spent, and conversion rate for how much time that translates to on the calendar. This is probably something we need to be explicit about when talking about estimates.

**Comment by slevis-lmwg** [comment]
@ekluzek I can track my c13/c14 Sprint hours in this issue, you can track yours in #3346 if that makes sense to you, too. Let me know if you disagree.

**Comment by slevis-lmwg** [comment]
@slevis-lmwg generate mesh for c14

**Comment by ekluzek** [comment]
Some discussion on the mesh issue is here: https://github.com/ESCOMP/CTSM/issues/3346#issuecomment-3393629862

**Comment by slevis-lmwg** [comment]
Regarding cmip6 stream files, [this post ](https://github.com/ESCOMP/CTSM/issues/3346#issuecomment-3398226829)lowers the priority at this time.

**Comment by slevis-lmwg** [comment]
@ekluzek I have completed all the steps and new files are in /glade/campaign/cesm/cesmdata/cseg/inputdata/lnd/clm2/isotopes:
```
ctsmforc.Graven.atm_delta_C13_CMIP7_global_1700-2023_yearly_v3.0_c251013.nc
ctsmforc.Graven.atm_delta_C14_CMIP7_4x1_global_1700-2023_yearly_v3.0_c251013.nc
mesh_4x1_global_c20251013.nc
```
- [x] ESMF_Scrip2Unstruct generated the mesh a little different from what you expected in [this post](https://github.com/ESCOMP/CTSM/issues/3346#issuecomment-3393629862), so we will see when you try it...
- [x] My script with all the steps is in `lnd/clm2/isotopes/rawdata_copies/workspace/prep_c13_14_streams.sh`, and it can go to the [cmip7 repo](https://github.com/NCAR/CMIP7_inputdata_processing) when we confirm that everything worked.

**Comment by ekluzek** [comment]
I think there might be some adjustments we should make to the datasets. I found an example of latitude bands that we can model after here:

$DIN_LOC_ROOT/atm/datm7/CO2/fco2_datm_lat-bandsSSP1-1.9_simyr_2014-2500_CMIP6_c190514.nc

There's some differences there that we might need to bring in place. It's also able to NOT have to have a separate mesh file which would be a good thing.

**Comment by slevis-lmwg** [comment]
From meeting with @ekluzek 
- We agreed to proceed with the files as they are for now, because we expect them to work.
- If they do not work, let's revisit the merge idea from the last post.
- If they do work, let's pursue the merge idea as nice-to-have in terms of reducing the number of files.

Also, I moved this to "stalled" until we know whether my work here needs modification. Else, this is ready to close.

**Comment by ekluzek** [comment]
Steps to complete:

- [x] b4b-dev for RC14_CANAIR #3548 
- [x] Answer changing tag on master to remove c13/c14 for SSP's #3549
- [x] Add new streams baseclass code, make sure it can compile / run (even though not called) b4b
- [x] Add the c13/c14 extensions b4b
- [x] Add read of streams which is always on, but not used b4b
- [x] Add new namelist items to CNCisoAtmReadMod b4b
- [x] Add new namelist items to the namelist_definition_ctsm.xml file b4b
- [x] #3561
- [x] Change Photosynthesis in currrent code so that sectors is internal to CNCIsoAtmReadMod rather than Photosynthesis b4b
- [x] Add option of reading in CIso from streams, with default off b4b
- [x] Show that streams results identical or almost identical to the previous option
- [x] Change default to read in from streams -- answer change

**Comment by ekluzek** [comment]
Ok it turns out I didn't appreciate some limitations on streams that make this harder than I thought. This comes about partially because the streams silently changes some of the options the user asks for.

The CDEPS parent issue for this is here:

https://github.com/ESCOMP/CDEPS/issues/358

My assumption at the beginning was that streams for mapping would work one of two ways. With a mesh file or the way the CO2 latitude bands works. I thought both were solid options. But it turns out they both have problems that weren't immediately apparent. Both can still be made to work, but more work than I expected.

The things that weren't true that I assumed were are:
- CO2 latitude bands was working correctly
- Mesh mapping could work for any size grid
- And if not only simple changes would be needed
- Streams would error check its inputs and die if something wasn't viable
- Streams wouldn't silently not do something the user asked for
- It would be easy to show that the CO2 latitude bands were correct
- That showing it was roughly right was sufficient
